lib/cast.js

```diff
@@ -305,38 +304,39 @@ module.exports = function cast(schema, obj, options, context) {
           let $cond;
 
           let k = ks.length;
-
           while (k--) {
             $cond = ks[k];
             nested = val[$cond];
-
-            if ($cond === '$not') {
-              if (nested && schematype) {
-                _keys = Object.keys(nested);
-                if (_keys.length && isOperator(_keys[0])) {
-                  for (const key in nested) {
-                    nested[key] = schematype.castForQuery(
-                      key,
-                      nested[key],
+            if (Object.keys(nested || {}).some(isOperator) && checkNested(nested)) {
+              cast(schema, nested, options, context);
```

Comment:
Maybe instead of casting the whole `nested` property, you instead loop through all the keys of `nested`, check if the key matches `isOperator`, and then cast that? That would save you from needing the `checkNested` function, which seems a bit unnecessary.

Reply:
@vkarpov15  Trying you're recommended approach seems to move the PR back to square 1 as it cause the same twenty test failures that the `checkNested()` function attempted to resolve. Unless I messed up your recommendation.

My implementation of the recommendation:
```javascript
if (Object.keys(nested || {}).some(isOperator)) {
              const iterate = (obj) => {
                Object.keys(obj).forEach(key => {
                  if (isOperator(key)) {
                    cast(schema, obj[key], options, context)
                  }
                })
              };
              iterate(nested);
            } 
```

Reply:
Upon further iteration the flow I have works because it catches the case where `nested = {$not: {$elemMatch: 'val'} }` and the like kicks it to the if statement on line 313